### PR TITLE
Turn on/off simulcast in config

### PR DESCRIPTION
**Pull request by Rados13**
This PR allows you to create a room where you can use simulcast.

To have a possibility to create a room with simulcast you must add query parameters simulcast example link for localhost is `http://localhost:4000/?simulcast`. On the page beside the standard fields `Room name` and `Display name,` there will also be a checkbox `Simulcast`. The first person who joins the room decides whether it will be with or without simulcast.

- In the right-down corner of simulcast, the room should be a button `Show simulcast controls`. 
- After clicking this button, controls on all video elements should show up. 
- On the video element of myself, it should display in the right-down corner a panel with the text `Encodings to send` and three checkboxes with labels `High`, `Medium`, and `Low`. 
- On the video element of other peers in the right-up corner should be the text `Encoding: <name of current encoding>`, in the right-down corner should show select element with prefix `Encoding to receive` and select should allow us to change between values `High`, `Medium`, `Low`. 
- After leaving the simulcast room, the query parameter `simulcast` should be filled in the link.

In the non-simulcast room, there shouldn't be a `Show simulcast controls` button or any controls regarding simulcast.

Currently, the priority of receiving encoding is like this: medium, high, and low. So by default, all peers should receive medium encoding from other peers.

Testing simulcast functionality is divided into two parts:
1. Changing encodings of myself:
    - after disabling medium encoding, other peers should start receiving high encoding of your video,
    - after disabling high encoding, other peers should start receiving low encoding
    - after disabling low encoding, other peers shouldn't receive any video of you
    - then after enabling encodings in any order, the encoding is forwarded to other peers accordingly to the priority described above.
2. Changing encoding of another peer:
    - after a few seconds from switching encoding to another encoding, you should start receiving a video at the new encoding, e.g if you switch to high you should start receiving video in resolution 1280x720.

### CODE REVIEW

**Comment by bblaszkow06** [comment]
* As discussed it would be better to use some other key for showing controls (or a button)
* It seems we might need to work on both versions for some time, so it would be best to have both variants available, e.g. by adding query parameter `?simulcast=true` or using a different URL (`simulcast/room/id`)

File: assets/src/room.ts [hunk 1]
```diff
@@ -277,11 +310,35 @@ export class Room {
...
+      )[0].trackId!;
+    this.webrtc.selectTrackEncoding(peerId, trackId, encoding);
+  };
+
   private parseUrl = (): string => {
     const { display_name: displayName } = parse(document.location.search);
 
     // remove query params without reloading the page
-    window.history.replaceState(null, "", window.location.pathname);
+    // window.history.replaceState(null, "", window.location.pathname);
```
Comment (mickel8) [inline assets/src/room.ts:341]: Is this needed?

File: assets/src/room_ui.ts [hunk 2]
```diff
@@ -8,24 +10,63 @@ const screensharingButton = document.getElementById(
...
+const simulcastControls = document.querySelector(
+  "div[name=simulcast-controls]"
+) as HTMLElement;
+const simulcastControlsBtn = simulcastControls.querySelector("button") as HTMLButtonElement;
+simulcastControlsBtn.onclick = () => {
+  state.showSimulcast = !state.showSimulcast
+  setShowSimulcast(state.showSimulcast);
+}
+simulcastControls.hidden = true
+simulcastControlsBtn.hidden = true
```
Comment (mickel8) [inline assets/src/room_ui.ts:22]: I think this should be in `setupControls`

File: lib/videoroom/room.ex [hunk 3]
```diff
@@ -49,6 +50,8 @@ defmodule Videoroom.Room do
         :error -> nil
       end
 
+    simulcast? = Application.fetch_env!(:membrane_videoroom_demo, :simulcast?)
```
Comment (mickel8) [inline lib/videoroom/room.ex:53]: As discussed let's move this on the client side and change query param into checkbox

**Review by mickel8 — CHANGES_REQUESTED** [review]

File: lib/videoroom_web/controllers/page_controller.ex [hunk 4]
```diff
@@ -5,13 +5,18 @@ defmodule VideoRoomWeb.PageController do
     render(conn, "index.html", room_id: Map.get(params, "room_id"))
   end
 
-  def enter(conn, %{"room_name" => room_name, "display_name" => display_name}) do
+  def enter(
+        conn,
+        %{"room_name" => room_name, "display_name" => display_name} = params
+      ) do
+    simulcast? = Map.get(params, "simulcast", false) == ""
```
Comment (mickel8) [inline lib/videoroom_web/controllers/page_controller.ex:12]: This should be always present, shouldn't it?

File: assets/src/room_ui.ts [hunk 5]
```diff
@@ -8,24 +10,58 @@ const screensharingButton = document.getElementById(
 const leaveButton = document.getElementById(
   "leave-control"
 ) as HTMLButtonElement;
+const simulcastControls = document.querySelector(
+  "div[name=simulcast-controls]"
+) as HTMLElement;
+const simulcastControlsBtn = simulcastControls.querySelector("button") as HTMLButtonElement;
```
Comment (mickel8) [inline assets/src/room_ui.ts:16]: Could we format `TS` format?

File: assets/src/room_ui.ts [hunk 6]
```diff
@@ -356,3 +463,43 @@ export function setErrorMessage(
...
+function changeVisibilitySimulcastControls(showSimulcast: boolean) {
+
+  let feeds = document.querySelectorAll(
+    "div[name=video-feed]"
+  ) as NodeListOf<HTMLElement>;
+
+  feeds.forEach((feed) => {
+    changeVisibilityLocalSimulcastControls(feed, !showSimulcast);
+
+    changeVisibilityRemoteSimulcastControls(feed, !showSimulcast);
```
Comment (mickel8) [inline assets/src/room_ui.ts:476]: Why do we negate `showSimulcast`?

File: assets/src/room_ui.ts [hunk 7]
```diff
@@ -57,6 +93,14 @@ export function setupControls(
     callbacks.onLeave();
     window.location.reload();
   };
+
+
+  simulcastControlsBtn.onclick = () => {
+    state.showSimulcast = !state.showSimulcast
+    setShowSimulcast(state.showSimulcast);
```
Comment (mickel8) [inline assets/src/room_ui.ts:100]: This button should be visible only when simulcast was enabled so we can do
```suggestion
    changeVisibilitySimulcastControlls(!state.showSimulcast)
```    

and delete `setShowSimulcast`

File: assets/src/room_ui.ts [hunk 8]
```diff
@@ -57,6 +93,14 @@ export function setupControls(
...
+  }
+  simulcastControls.hidden = true
```
Comment (mickel8) [inline assets/src/room_ui.ts:102]:
```suggestion
  simulcastControlsBtnDiv.hidden = true
```
Also, do we have to hide button when we have already hide the whole div?

**Review by mickel8 — APPROVED** [review]

**Review by mickel8 — APPROVED** [review]

File: lib/videoroom_web/controllers/room_controller.ex [hunk 9]
```diff
@@ -1,13 +1,18 @@
 defmodule VideoRoomWeb.RoomController do
   use VideoRoomWeb, :controller
 
-  def index(conn, %{"room_id" => id, "display_name" => _name}) do
+  def index(conn, %{"room_id" => id, "display_name" => _name} = params) do
```
Comment (bblaszkow06) [inline lib/videoroom_web/controllers/room_controller.ex:4]: Warning: unused variable

**Review by bblaszkow06 — CHANGES_REQUESTED** [review]

**Review by bblaszkow06 — APPROVED** [review]

**Review by maciejewskakatarzyna — CHANGES_REQUESTED** [review]
## 3 bugs below:

### BUG 1
**Description**: 
Mobile: bottom controls are squeezed and simulcast controls cover up the video (screen)

**Steps to reproduce:**
Create a room on mobile with at least one person and simulcast turned on.
Click "Show simulcast controls" button.

**Expected situation:** 
Videoroom controls should be not squeezed (circle) and visible. 
Simulcast controls and video should be visible.

**Environment:** 
iOS: Chrome, Safari (iPhone 12)
Android: Chrome (Samsung A53)

IMG_3895

----------

### BUG 2
**Description**: 
Mobile: simulcast controls for another peer are not visible (screen)

**Steps to reproduce:**
Create a room on mobile with at least two people and simulcast turned on.
Click "Show simulcast controls" button.

**Expected situation:** 
Simulcast controls should be visible for both users.

**Environment:** 
iOS: Chrome, Safari (iPhone 12)
Android: Chrome (Samsung A53)

IMG_3896

-----------

### BUG 3
**Description**: 
Simulcast controls are visible after joining the room with simulcast checkbox unchecked (video)

**Steps to reproduce:**
Create a room with at least one person and simulcast turned on.
Leave the room.
Uncheck simulcast checkbox.
Join the room.
Click "Show simulcast controls" button.

**Expected situation:** 
Simulcast controls shouldn't be visible after joining the room with simulcast checkbox unchecked.

**Environment:** 
macOS: Chrome, Safari
iOS: Chrome, Safari (iPhone 12)
Android: Chrome (Samsung A53)

https://user-images.githubusercontent.com/52798715/184877372-ef679afc-f176-450c-9985-e04b5d89cef0.mov

File: lib/videoroom_web/templates/room/index.html.heex [hunk 10]
```diff
@@ -33,6 +41,15 @@
                     <img class="invisible invert" src={static_path(@conn, "/svg/mic-off-fill.svg")} style="position: absolute; left: 0; top: 0; width: 10%, height: 10%">
                     <div name="video-label" class="absolute text-white text-shadow-lg bottom-0 left-0 p-2">Placeholder</div>
                     <div name="no-video-msg" class="absolute text-white w-full h-full top-0 left-0" style="text-align: center">Camera turned off</div>
+                    <div name="video-encoding", class="absolute text-green-700/100 text-shadow-lg bg-white top-0 right-0 p-2", style="color: green, font-size: 20px">Encoding: m</div>
```
[the remark this answers is not in the document]
Comment (maciejewskakatarzyna) [reply lib/videoroom_web/templates/room/index.html.heex:44]: FIXED

**Review by maciejewskakatarzyna — APPROVED** [review]

**Pull request merged**